File: apps/web/src/services/recoveryChatSink.ts
import { RecoveryNotificationJournal, type NotificationJournalImage } from '../../../../packages/domain-ts/src/notificationRecovery'
import type { ChatMessage, OutboxEntry } from '../../../../packages/domain-ts/src/models'
import { upsertOutbox } from '../../../../packages/domain-ts/src/outbox'
import { applyRecoveryData } from '../../../../packages/domain-ts/src/recoveryData'
import { initialRecoveryState, recoveryDecimal, sameRecoveryContext, RecoveryProtocolError,
  type RecoveryContext, type RecoveryPlan, type RecoveryState } from '../../../../packages/domain-ts/src/recovery'
import type { SnapshotItem } from '../../../../packages/domain-ts/src/recoverySnapshot'
import type { RecoverySink } from './recoveryCoordinator'
import { loadRecoverySnapshot, commitRecoverySnapshot } from './localChatDb'

export interface RecoveredChatImage {
  notificationRecovery: NotificationJournalImage
  messages: ChatMessage[]
  outbox: OutboxEntry[]
  directory: SnapshotItem[]
  positions: Record<string, number>
  tombstones: SnapshotItem[]
}
export interface RecoveryChatStorage {
  loadRecoverySnapshot: typeof loadRecoverySnapshot
  commitRecoverySnapshot: typeof commitRecoverySnapshot
}
const clone = <T>(value: T): T => JSON.parse(JSON.stringify(value)) as T

/** Rebuilds existing message/outbox models; directory proof is not a fabricated friend/group profile. */
export class RecoveryChatSink implements RecoverySink {
  private context?: RecoveryContext
  private active = false
  private revision = 0
  private run = 0
  private committedCursor: string | null = null
  private notifications = new RecoveryNotificationJournal()
  private messages = new Map<string, ChatMessage>()
  private outbox: OutboxEntry[] = []
  private directory = new Map<string, SnapshotItem>()
  private positions: Record<string, number> = {}
  private tombstones = new Map<string, SnapshotItem>()
  constructor(private readonly seed: () => {outbox: readonly OutboxEntry[]},
    private readonly isCurrent: (context: RecoveryContext) => boolean,
    private readonly onPublish: (state: RecoveryState, image: RecoveredChatImage) => void,
    private readonly onQuarantine: (reason: string) => void,
    private readonly storage: RecoveryChatStorage = {loadRecoverySnapshot, commitRecoverySnapshot},
    private readonly liveCandidates:()=>readonly string[]=()=>[]) {}
  private check(context: RecoveryContext): void {
    if (!this.active || !this.context || !sameRecoveryContext(this.context,context) || !this.isCurrent(context))
      throw new RecoveryProtocolError('STALE_RECOVERY')
  }
  private owner(context: RecoveryContext): string {return `${context.origin}|${context.userId}`}
  private sequence(raw: unknown): number {
    const value = recoveryDecimal(raw)
    if(value > BigInt(Number.MAX_SAFE_INTEGER)) throw new RecoveryProtocolError('UNSUPPORTED_SECURE_STATE')
    return Number(value)
  }
  async begin(context: RecoveryContext): Promise<void> {
    const run=++this.run
    this.context={...context}; this.active=true; this.committedCursor=null; this.check(context)
    this.messages.clear();this.directory.clear();this.positions={};this.tombstones.clear()
    const seeded=clone(this.seed().outbox)
    const stored=await this.storage.loadRecoverySnapshot(this.owner(context));this.check(context)
    if(run!==this.run) throw new RecoveryProtocolError('STALE_RECOVERY')
    this.notifications=new RecoveryNotificationJournal(stored?.snapshot.notificationRecovery)
    let pending: OutboxEntry[]=[]
    if(stored) {
      if(!Array.isArray(stored.snapshot.outbox)) throw new RecoveryProtocolError('UNSUPPORTED_SECURE_STATE')
      for(const raw of stored.snapshot.outbox) {
        if(!raw || typeof raw!=='object' || typeof raw.clientMsgId!=='string' || !raw.clientMsgId
          || typeof raw.conversationId!=='string' || typeof raw.requestId!=='string'
          || typeof raw.createdAt!=='string' || !Number.isSafeInteger(raw.retryCount) || raw.retryCount<0
          || !['WAITING_NETWORK','SENDING','FAILED'].includes(raw.state)
          || (raw.recoveryDisposition!==undefined && !['NEEDS_USER_ACTION','DROP_BODY_REVOKED'].includes(raw.recoveryDisposition))
          || !raw.payload || typeof raw.payload.content!=='string' || typeof raw.payload.contentType!=='string')
          throw new RecoveryProtocolError('UNSUPPORTED_SECURE_STATE')
        pending=upsertOutbox(pending,clone(raw as OutboxEntry))
      }
    }
    for(const entry of seeded) pending=upsertOutbox(pending,entry)
    this.outbox=applyRecoveryData({next:initialRecoveryState(context),changed:true,ignored:false,
      effects:{eraseMessages:new Set(),revokeConversations:new Set(),rebuildConversations:new Set(),
        stopAutomaticSend:new Set(pending.map(e=>e.conversationId))}},[],pending).outbox
    this.revision=stored?.revision ?? 0
  }
  async resume(context: RecoveryContext, state: RecoveryState): Promise<void> {
    if(!this.context || !sameRecoveryContext(this.context,context) || !sameRecoveryContext(state.context,context) || !this.isCurrent(context))
      throw new RecoveryProtocolError('STALE_RECOVERY')
    this.run++;this.active=true;this.committedCursor=null
  }
  async snapshot(context: RecoveryContext, items: readonly SnapshotItem[]): Promise<void> {
    this.check(context);this.committedCursor=null
    const run=this.run
    const terminal: SnapshotItem[]=[]
    for(const item of items) {
      this.check(context)
      if(run!==this.run) throw new RecoveryProtocolError('STALE_RECOVERY')
      const cid=item.conversationId
      if(item.kind==='CONVERSATION') {
        if(!item.readAllowed) {
          await this.mutations(context,{next:initialRecoveryState(context),changed:true,ignored:false,
            effects:{eraseMessages:new Set(),revokeConversations:new Set([cid]),stopAutomaticSend:new Set([cid]),rebuildConversations:new Set()}})
        } else {this.directory.set(cid,clone(item));this.positions[cid]=this.sequence(item.messageSequenceAtH)}
      } else if(item.state==='NORMAL') {
        const details=item.details!
        const message: ChatMessage={...details,messageId:item.messageId!,conversationId:cid,
          fromUserId:Number(details.fromUserId),type:String(details.contentType),content:item.content,
          sequence:this.sequence(item.messageSequence),deliveryState:'SENT'}
        this.messages.set(message.messageId,message)
      } else {
        terminal.push(item)
      }
    }
    if(terminal.length) {
      const state=initialRecoveryState(context)
      await this.mutations(context,{next:{...state,messages:new Map(terminal.map(item=>[item.messageId!,{
        conversationId:item.conversationId,objectVersion:item.objectVersion!,state:item.state!}]))},changed:true,ignored:false,
        effects:{eraseMessages:new Set(terminal.map(item=>item.messageId!)),revokeConversations:new Set(),stopAutomaticSend:new Set(),rebuildConversations:new Set()}})
      this.check(context)
      if(run!==this.run) throw new RecoveryProtocolError('STALE_RECOVERY')
      for(const item of terminal) this.tombstones.set(item.messageId!,clone(item))
    }
  }

  async mutations(context: RecoveryContext, plan: RecoveryPlan): Promise<void> {
    this.check(context);this.committedCursor=null
    if(!sameRecoveryContext(context,plan.next.context)) throw new RecoveryProtocolError('STALE_RECOVERY')
    const changed=applyRecoveryData(plan,[...this.messages.values()],this.outbox)
    this.messages=new Map(changed.messages.map(message=>[message.messageId,message]));this.outbox=changed.outbox
    for(const cid of plan.effects.revokeConversations) {
      this.directory.delete(cid);delete this.positions[cid]
      for(const [id,item] of this.tombstones) if(item.conversationId===cid) this.tombstones.delete(id)
    }
    for(const id of plan.effects.eraseMessages) {
      const proof=plan.next.messages.get(id)!
      const sequence=this.messages.get(id)?.sequence
      this.tombstones.set(id,{kind:'MESSAGE',conversationId:proof.conversationId,messageId:id,
        objectVersion:proof.objectVersion,state:proof.state,
        ...(sequence ? {messageSequence:String(sequence)} : this.tombstones.get(id)?.messageSequence
          ? {messageSequence:this.tombstones.get(id)!.messageSequence} : {})})
    }
  }
  private image(state: RecoveryState): RecoveredChatImage {
    return clone({notificationRecovery:this.notifications.image(),messages:[...this.messages.values()].filter(m=>state.access.get(m.conversationId || '')?.readAllowed
      && state.messages.get(m.messageId)?.state==='NORMAL'),outbox:this.outbox,
      directory:[...this.directory.values()],positions:this.positions,tombstones:[...this.tombstones.values()]})
  }
  planLiveNotifications(context: RecoveryContext,state: RecoveryState,ids:readonly string[]):void {
    this.check(context);this.committedCursor=null;this.notifications.reconcile(state,ids)
  }
  acknowledgeNotification(context:RecoveryContext,key:string):void {
    this.check(context);this.committedCursor=null;this.notifications.acknowledge(key)
  }
  replaceOutbox(context: RecoveryContext, entries: readonly OutboxEntry[]): void {
    this.check(context);this.committedCursor=null
    this.outbox=clone([...entries])
  }
  async commit(context: RecoveryContext, state: RecoveryState): Promise<void> {
    this.check(context);this.committedCursor=null
    if(!sameRecoveryContext(context,state.context)) throw new RecoveryProtocolError('STALE_RECOVERY')
    const run=this.run
    const absent=new Set(this.outbox.filter(e=>!state.access.get(e.conversationId)?.readAllowed).map(e=>e.conversationId))
    await this.mutations(context,{next:state,changed:absent.size>0,ignored:false,
      effects:{eraseMessages:new Set(),revokeConversations:absent,stopAutomaticSend:absent,rebuildConversations:new Set()}})
    this.check(context)
    if(run!==this.run) throw new RecoveryProtocolError('STALE_RECOVERY')
    this.notifications.reconcile(state,this.liveCandidates().filter(id=>{const message=this.messages.get(id);return message && String(message.fromUserId)!==context.userId}))
    const snapshot={...this.image(state),recovery:{origin:context.origin,userId:context.userId,streamEpoch:context.streamEpoch,
      cursor:state.cursor,messages:Object.fromEntries(state.messages),access:Object.fromEntries(state.access),
      seen:Object.fromEntries(state.seen),rebuild:[...state.rebuild]}}
    const revision=await this.storage.commitRecoverySnapshot(this.owner(context),this.revision,snapshot)
    this.check(context)
    if(run!==this.run) throw new RecoveryProtocolError('STALE_RECOVERY')
    this.revision=revision;this.committedCursor=state.cursor
  }
  publish(context: RecoveryContext, state: RecoveryState): void {
    this.check(context)
    if(!sameRecoveryContext(context,state.context) || state.phase!=='ONLINE_SAFE' || state.cursor!==this.committedCursor)
      throw new RecoveryProtocolError('PROTOCOL_ERROR')
    this.onPublish(state,this.image(state))
  }
  quarantine(reason: string): void {this.run++;this.active=false;this.committedCursor=null;this.onQuarantine(reason)}
}

File: apps/flutter-prototype/lib/application/recovery_chat_sink.dart
import '../core/json_views.dart';
import '../core/models.dart';
import '../core/notification_recovery.dart';
import '../core/store.dart';
import '../core/recovery.dart';
import '../core/recovery_data.dart';
import '../core/recovery_snapshot.dart';
import 'recovery_coordinator.dart';

// Top-level serializers avoid sending controller closures with isolated storage.
Json _encodeChatMessage(ChatMessage message) => message.toJson();
Json _encodeMessageProof(MessageRecoveryProof proof) => {
  'conversationId': proof.conversationId,
  'objectVersion': proof.objectVersion,
  'state': proof.state.name.toUpperCase(),
};

/// Existing chat models plus body-free terminal headers; no invented sender/date for tombstones.
class RecoveredChatImage {
  RecoveredChatImage(
    this.chat,
    this.outbox,
    this.tombstones,
    this.notificationRecovery,
  );
  final Json notificationRecovery;
  final Map<String, List<ChatMessage>> outbox;
  final RecoveryChatData chat;
  final Map<String, SnapshotItem> tombstones;
}

class RecoveryChatSink implements RecoverySink {
  RecoveryChatSink({
    required this.store,
    required this.seed,
    required this.isCurrent,
    required this.onPublish,
    required this.onQuarantine,
    this.liveCandidates,
  });
  final List<String> Function()? liveCandidates;
  final RecoveryChatStore store;
  final RecoveryChatData Function() seed;
  final bool Function(RecoveryContext) isCurrent;
  final void Function(RecoveryState, RecoveredChatImage) onPublish;
  final void Function(String) onQuarantine;
  RecoveryContext? _context;
  int _revision = 0, _run = 0;
  bool _active = false;
  String? _committedCursor;
  RecoveryChatData? _chat;
  RecoveryNotificationJournal _notifications = RecoveryNotificationJournal();
  final Map<String, SnapshotItem> _tombstones = {};
  final Map<String, String> _titles = {};
  String _owner(RecoveryContext context) =>
      '${context.origin}|${context.userId}';
  void _check(RecoveryContext context) {
    if (!_active || _context?.matches(context) != true || !isCurrent(context)) {
      throw const RecoveryProtocolError('STALE_RECOVERY');
    }
  }

  int _sequence(Object? value) {
    final n = recoveryDecimal(value);
    if (n > BigInt.from(9007199254740991)) {
      throw const RecoveryProtocolError('UNSUPPORTED_SECURE_STATE');
    }
    return n.toInt();
  }

  Conversation _conversation(String cid, RecoveryContext context) {
    final parts = cid.split(':');
    String kind;
    int target;
    if (parts.length == 3 &&
        parts[0] == 'private' &&
        (parts[1] == context.userId || parts[2] == context.userId)) {
      kind = 'private';
      target = _sequence(parts[1] == context.userId ? parts[2] : parts[1]);
    } else if (parts.length == 2 && {'group', 'temporary'}.contains(parts[0])) {
      kind = parts[0];
      target = _sequence(parts[1]);
    } else {
      throw const RecoveryProtocolError('PROTOCOL_ERROR');
    }
    if (target <= 0) throw const RecoveryProtocolError('PROTOCOL_ERROR');
    return Conversation(
      id: cid,
      targetId: target,
      kind: kind,
      title: _titles[cid] ?? '会话 $target',
    );
  }

  @override
  Future<void> begin(RecoveryContext context) async {
    final run = ++_run;
    _context = context;
    _active = true;
    _committedCursor = null;
    _check(context);
    final old = seed();
    _titles.clear();
    _tombstones.clear();
    for (final conversation in old.conversations) {
      _titles[conversation.id] = conversation.title;
    }
    final pending = <String, List<ChatMessage>>{};
    for (final entry in old.messages.entries) {
      final held = entry.value
          .where(
            (message) =>
                message.fromUserId.toString() == context.userId &&
                (message.delivery != Delivery.sent ||
                    message.recoveryDisposition != null),
          )
          .map(
            (message) => message.withRecoveryDisposition(
              RecoveryDisposition.needsUserAction,
            ),
          )
          .toList();
      if (held.isNotEmpty) pending[entry.key] = held;
    }
    _chat = RecoveryChatData(pending, [], {});
    final existing = await store.loadRecovery(_owner(context));
    _check(context);
    if (run != _run) {
      throw const RecoveryProtocolError('STALE_RECOVERY');
    }
    _revision = existing?.revision ?? 0;
    _notifications = RecoveryNotificationJournal(
      existing?.snapshot['notificationRecovery'],
    );
    if (existing != null) {
      final saved = existing.snapshot['outbox'];
      if (saved is! Map) {
        throw const RecoveryProtocolError('UNSUPPORTED_SECURE_STATE');
      }
      for (final entry in saved.entries) {
        if (entry.key is! String || entry.value is! List) {
          throw const RecoveryProtocolError('UNSUPPORTED_SECURE_STATE');
        }
        final rows = (entry.value as List).map((raw) {
          final message = ChatMessage.restore(
            Map<String, dynamic>.from(raw as Map),
          );
          if (message.conversationId != entry.key ||
              message.fromUserId.toString() != context.userId ||
              message.delivery == Delivery.sent) {
            throw const RecoveryProtocolError('UNSUPPORTED_SECURE_STATE');
          }
          return message.withRecoveryDisposition(
            RecoveryDisposition.needsUserAction,
          );
        }).toList();
        pending[entry.key as String] = mergeMessages(
          rows,
          pending[entry.key] ?? [],
        );
      }
    }
  }

  @override
  Future<void> resume(RecoveryContext context, RecoveryState state) async {
    if (_context?.matches(context) != true ||
        !state.context.matches(context) ||
        _chat == null ||
        !isCurrent(context)) {
      throw const RecoveryProtocolError('STALE_RECOVERY');
    }
    ++_run;
    _active = true;
    _committedCursor = null;
  }

  @override
  Future<void> snapshot(
    RecoveryContext context,
    List<SnapshotItem> items,
  ) async {
    _check(context);
    _committedCursor = null;
    var chat = _chat!;
    for (final item in items) {
      final cid = item['conversationId'] as String;
      if (item['kind'] == 'CONVERSATION') {
        if (item['readAllowed'] == false) {
          final plan = RecoveryPlan(
            RecoveryState(context: context),
            RecoveryEffects(
              revokeConversations: {cid},
              stopAutomaticSend: {cid},
            ),
            changed: true,
          );
          chat = applyRecoveryChatData(plan, chat);
          chat = RecoveryChatData(
            chat.messages,
            chat.conversations.where((c) => c.id != cid).toList(),
            chat.positions,
          );
          _tombstones.removeWhere((_, value) => value['conversationId'] == cid);
        } else {
          final conversations = [
            ...chat.conversations.where((c) => c.id != cid),
            _conversation(cid, context),
          ];
          chat = RecoveryChatData(chat.messages, conversations, {
            ...chat.positions,
            cid: _sequence(item['messageSequenceAtH']),
          });
        }
      } else if (item['state'] == 'NORMAL') {
        final details = Map<String, dynamic>.from(item['details'] as Map);
        final message = ChatMessage.fromJson({
          ...details,
          'messageId': item['messageId'],
          'conversationId': cid,
          'content': item['content'],
          'sequence': _sequence(item['messageSequence']),
        });
        final rows = [...chat.messages[cid] ?? <ChatMessage>[]];
        rows.removeWhere(
          (old) =>
              old.delivery == Delivery.sent &&
              old.messageId == message.messageId,
        );
        rows.add(message);
        chat = RecoveryChatData(
          {...chat.messages, cid: rows},
          chat.conversations,
          chat.positions,
        );
      } else {
        final id = item['messageId'] as String;
        _tombstones[id] = {...item};
        final proof = MessageRecoveryProof(
          cid,
          item['objectVersion'] as String,
          MessageRecoveryState.values.byName(
            (item['state'] as String).toLowerCase(),
          ),
        );
        chat = applyRecoveryChatData(
          RecoveryPlan(
            RecoveryState(context: context, messages: {id: proof}),
            RecoveryEffects(eraseMessages: {id}),
            changed: true,
          ),
          chat,
        );
        chat = RecoveryChatData(
          {
            ...chat.messages,
            cid: (chat.messages[cid] ?? [])
                .where((m) => m.messageId != id || m.delivery != Delivery.sent)
                .toList(),
          },
          chat.conversations,
          chat.positions,
        );
      }
    }
    _chat = chat;
  }

  @override
  Future<void> mutations(RecoveryContext context, RecoveryPlan plan) async {
    _check(context);
    _committedCursor = null;
    _chat = applyRecoveryChatData(plan, _chat!);
    for (final cid in plan.effects.revokeConversations) {
      _chat = RecoveryChatData(
        _chat!.messages,
        _chat!.conversations.where((c) => c.id != cid).toList(),
        _chat!.positions,
      );
      _tombstones.removeWhere((_, value) => value['conversationId'] == cid);
    }
    for (final id in plan.effects.eraseMessages) {
      final proof = plan.next.messages[id]!;
      final rows = _chat!.messages[proof.conversationId] ?? [];
      final matches = rows.where((m) => m.messageId == id && m.sequence > 0);
      final old = _tombstones[id];
      _tombstones[id] = {
        'kind': 'MESSAGE',
        'conversationId': proof.conversationId,
        'messageId': id,
        'objectVersion': proof.objectVersion,
        'state': proof.state.name.toUpperCase(),
        if (matches.isNotEmpty)
          'messageSequence': '${matches.first.sequence}'
        else if (old?['messageSequence'] != null)
          'messageSequence': old!['messageSequence'],
      };
    }
  }

  void planLiveNotifications(
    RecoveryContext context,
    RecoveryState state,
    List<String> ids,
  ) {
    _check(context);
    _committedCursor = null;
    _notifications.reconcile(state, ids);
  }

  void acknowledgeNotification(RecoveryContext context, String key) {
    _check(context);
    _committedCursor = null;
    _notifications.acknowledge(key);
  }

  /// User-created pending rows are persisted with the already verified history.
  /// Incoming WS/history bodies must never enter through this method.
  void replacePending(
    RecoveryContext context,
    Map<String, List<ChatMessage>> rows,
  ) {
    _check(context);
    _committedCursor = null;
    final combined = <String, List<ChatMessage>>{
      for (final entry in _chat!.messages.entries)
        entry.key: entry.value
            .where((m) => m.delivery == Delivery.sent)
            .toList(),
    };
    for (final entry in rows.entries) {
      for (final message in entry.value) {
        if (message.delivery == Delivery.sent) continue;
        if (message.fromUserId.toString() != context.userId ||
            message.conversationId != entry.key) {
          throw const RecoveryProtocolError('STALE_RECOVERY');
        }
        combined.putIfAbsent(entry.key, () => []).add(message);
      }
    }
    _chat = RecoveryChatData(combined, _chat!.conversations, _chat!.positions);
  }

  @override
  Future<void> commit(RecoveryContext context, RecoveryState state) async {
    _check(context);
    if (!state.context.matches(context)) {
      throw const RecoveryProtocolError('STALE_RECOVERY');
    }
    _committedCursor = null;
    final absent = _chat!.messages.keys
        .where((cid) => state.access[cid]?.readAllowed != true)
        .toSet();
    _chat = applyRecoveryChatData(
      RecoveryPlan(
        state,
        RecoveryEffects(revokeConversations: absent, stopAutomaticSend: absent),
        changed: absent.isNotEmpty,
      ),
      _chat!,
    );
    final liveIds = (liveCandidates?.call() ?? <String>[])
        .where(
          (id) => _chat!.messages.values
              .expand((rows) => rows)
              .any(
                (m) =>
                    m.messageId == id &&
                    m.fromUserId.toString() != context.userId,
              ),
        )
        .toList();
    _notifications.reconcile(state, liveIds);
    final image = _image(state);
    final chat = image.chat;
    final snapshot = <String, dynamic>{
      'conversations': chat.conversations.map((c) => c.toJson()).toList(),
      'messages': {
        for (final entry in chat.messages.entries)
          entry.key: JsonListView(entry.value, _encodeChatMessage),
      },
      'outbox': {
        for (final entry in image.outbox.entries)
          entry.key: JsonListView(entry.value, _encodeChatMessage),
      },
      'positions': chat.positions,
      'notificationRecovery': _notifications.image(),
      'tombstones': _tombstones,
      'recovery': {
        'origin': context.origin,
        'userId': context.userId,
        'streamEpoch': context.streamEpoch,
        'cursor': state.cursor,
        'messages': JsonMapView(state.messages, _encodeMessageProof),
        'access': {
          for (final entry in state.access.entries)
            entry.key: {
              'accessVersion': entry.value.accessVersion,
              'readAllowed': entry.value.readAllowed,
              'sendAllowed': entry.value.sendAllowed,
            },
        },
        'seen': state.seen,
        'rebuild': state.rebuild.toList(),
      },
    };
    final run = _run;
    final revision = await store.commitRecovery(
      _owner(context),
      _revision,
      snapshot,
    );
    _check(context);
    if (run != _run) {
      throw const RecoveryProtocolError('STALE_RECOVERY');
    }
    _revision = revision;
    _committedCursor = state.cursor;
  }

  @override
  void publish(RecoveryContext context, RecoveryState state) {
    _check(context);
    if (!state.context.matches(context) ||
        state.phase != RecoveryPhase.onlineSafe ||
        state.cursor != _committedCursor) {
      throw const RecoveryProtocolError('PROTOCOL_ERROR');
    }
    onPublish(state, _image(state));
  }

  RecoveredChatImage _image(RecoveryState state) {
    final history = <String, List<ChatMessage>>{};
    final outbox = <String, List<ChatMessage>>{};
    for (final entry in _chat!.messages.entries) {
      history[entry.key] = entry.value
          .where(
            (m) =>
                m.delivery == Delivery.sent &&
                state.access[m.conversationId]?.readAllowed == true &&
                state.messages[m.messageId]?.state ==
                    MessageRecoveryState.normal,
          )
          .toList();
      outbox[entry.key] = entry.value
          .where((m) => m.delivery != Delivery.sent)
          .toList();
    }
    return RecoveredChatImage(
      RecoveryChatData(history, _chat!.conversations, _chat!.positions),
      outbox,
      Map.unmodifiable(_tombstones),
      _notifications.image(),
    );
  }

  @override
  void quarantine(String reason) {
    ++_run;
    _active = false;
    _committedCursor = null;
    onQuarantine(reason);
  }
}
